strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -134,6 +350,319 @@ def prepare_gaussian_state(self, r, V, modes):
+
+        # Creating and calculating the weigths array for oscillating terms
+        odd_terms = np.array(range(-2 * z_max, 2 * z_max + 1), dtype=int) % 2
+        even_terms = (odd_terms + 1) % 2
+        even_phases = (-1) ** ((np.array(range(-2 * z_max, 2 * z_max + 1), dtype=int) % 4) // 2)
+        odd_phases = (-1) ** (
+            1 + ((np.array(range(-2 * z_max, 2 * z_max + 1), dtype=int) + 2) % 4) // 2
+        )
+        weights = np.cos(phi) * even_terms * even_phases * np.exp(
+            -0.5 * coef_sigma * p_means ** 2
+        ) - np.sin(phi) * odd_terms * odd_phases * np.exp(-0.5 * coef_sigma * p_means ** 2)
+        weights *= prefac
+        weights_real = np.ones(2, dtype=float)
+        weights = norm * np.concatenate((weights_real, weights))
+
+        # making sure the state is properly normalized
+        weights /= np.sum(weights)
+
+        # computing the means array
+        means = np.concatenate(
+            (
+                np.reshape(x_means, (-1, 1)),
+                np.reshape(p_means, (-1, 1)),
+            ),
+            axis=1,
+        )
+        means *= num_mean / denom_mean
+        means_real = np.sqrt(2 * hbar) * np.array([[a, 0], [-a, 0]], dtype=float)
+        means = np.concatenate((means_real, means))
+
+        # computing the covariance array
+        cov = np.array([[0.5 * hbar, 0], [0, (E * v) / (E + v)]])
+        cov = np.repeat(cov[None, :], 4 * z_max + 1, axis=0)
+        cov_real = 0.5 * hbar * np.array([[[1, 0], [0, 1]], [[1, 0], [0, 1]]], dtype=float)
+        cov = np.concatenate((cov_real, cov))
+
+        # filter out 0 components
+        filt = ~np.isclose(weights, 0, atol=cutoff)
+        weights = weights[filt]
+        means = means[filt]
+        cov = cov[filt]
+
+        # applying a rotation if necessary
+        if not np.isclose(phase, 0):
+            S = np.array([[np.cos(phase), -np.sin(phase)], [np.sin(phase), np.cos(phase)]])
+            means = np.dot(S, means.T).T
+            cov = S @ cov @ S.T
+
+        return weights, means, cov
+
+    def prepare_gkp(self, state, epsilon, cutoff, desc="real", shape="square"):
+        r"""Prepares the arrays of weights, means and covs for a finite energy GKP state.
+
+        GKP states are qubits, with the qubit state defined by:
+
+        :math:`\ket{\psi}_{gkp} = \cos\frac{\theta}{2}\ket{0}_{gkp} + e^(-i\phi)\sin\frac{\theta}{2}\ket{1}_{gkp}`
+
+        where the computational basis states are :math:`\ket{\mu}_{gkp} = \sum_{n} \ket{(2n+\mu)\sqrt{\pi\hbar}}_{q}`.
+
+        Args:
+            state (list): [theta,phi] for qubit definition above
+            epsilon (float): finite energy parameter of the state
+            cutoff (float): if using the 'real' representation, this determines
+                how many terms to keep
+            desc (string): 'real' or 'complex' reprsentation
+            shape (string): 'square' lattice or otherwise
```

Comment:
What do you mean by this? Can it be any shape or must it be a square lattice? Maybe better have something like "shape (string): shape of the lattice; default 'square'".

Reply:
Also, (minor) I think we more often use `(str)` instead of `(string)` in the code base, although I'm not sure what's actually preferred, or if it matters much.